### PR TITLE
[20.05] Invalidate tool cache when tool modified outside of Galaxy

### LINKED ISSUE #9863: Caching issue when rebuilding
When updating a tool xml file and rerunning the app, it seems we are not invalidating the old cache correctly, as updates are not applied to the running instance.

### PR DESCRIPTION
**Pull request by mvdbeek**
We store modtime and if the modtime changes we regenerate the cache entry. Fixes https://github.com/galaxyproject/galaxy/issues/9863

### CODE REVIEW

File: lib/galaxy/tool_util/parser/xml.py [hunk 1]
```diff
@@ -502,6 +504,13 @@ def parse_help(self):
     def macro_paths(self):
         return self._macro_paths
 
+    def paths_and_modtimes(self):
+        if not self._paths_and_modtimes:
+            self._paths_and_modtimes = {p: os.path.getmtime(p) for p in self._macro_paths}
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/parser/xml.py]:
```suggestion
            self._paths_and_modtimes = {p: os.path.getmtime(p) for p in self.macro_paths()}
```

File: lib/galaxy/tool_util/parser/xml.py [hunk 2]
```diff
@@ -502,6 +504,13 @@ def parse_help(self):
...
+            if self._source_path:
+                self._paths_and_modtimes[self._source_path] = os.path.getmtime(self._source_path)
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/parser/xml.py]:
```suggestion
            source_path = getattr(self, "_source_path", None)
            if source_path:
                self._paths_and_modtimes[source_path] = os.path.getmtime(source_path)
```

**Review by nsoranzo — COMMENTED** [review]
I think with my proposed changes you can move the whole implementation to `lib/galaxy/tool_util/parser/interface.py`

File: lib/galaxy/tool_util/parser/xml.py [hunk 3]
```diff
@@ -502,6 +504,13 @@ def parse_help(self):
     def macro_paths(self):
         return self._macro_paths
 
+    def paths_and_modtimes(self):
+        if not self._paths_and_modtimes:
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/parser/xml.py]: Is it safe to store this in the `XmlToolSource` object or is it better to regenerate the return value at each call?

**Comment by mvdbeek** [comment]
> I think with my proposed changes you can move the whole implementation to `lib/galaxy/tool_util/parser/interface.py`

`if getattr(instance, "attribute", default)` always feels quite hacky to me ... and I think it makes sense to separate how you get `paths_and_modtimes` for the different ToolSource classes, since I don't know what macro mechanism we're going to use in YamlToolSource and the other variants.

File: lib/galaxy/tool_util/parser/xml.py (hunk 3 above)
Comment (mvdbeek) [reply lib/galaxy/tool_util/parser/xml.py]: We're not re-using it right now, so yes, we can do that.

**Comment by nsoranzo** [comment]
> `if hasattr(instance, "attribute")` always feels quite hacky to me ... and I think it makes sense to separate how you get `paths_and_modtimes` for the different ToolSource classes, since I don't know what macro mechanism we're going to use in YamlToolSource and the other variants.

We can simple define a `source_path` property like we have `macro_paths` (which also could be a property BTW). All the 3 present subclasses of `ToolSource` have `self._source_path` defined in their `__init__()` method, so we don't need 3 different implementations.

File: lib/galaxy/tool_util/parser/xml.py [hunk 4]
```diff
@@ -499,9 +499,14 @@ def parse_help(self):
         help_elem = self.root.find('help')
         return help_elem.text if help_elem is not None else None
 
+    @property
     def macro_paths(self):
         return self._macro_paths
 
+    @property
+    def source_path(self):
+        return self._source_path
```
Comment (nsoranzo) [inline lib/galaxy/tool_util/parser/xml.py:508]: Can this be the default implementation in `ToolSource()` ? Or copy this to `CwlToolSource` and `YamlToolSource`.

**Review by nsoranzo — APPROVED** [review]

**Review by assuntad23 — APPROVED** [review]
This fixed it! Thanks!

**Pull request merged**